lib/modules/manager/bazel-module/fragments.ts

```diff
@@ -0,0 +1,269 @@
+}
+
+export class BooleanFragment implements FragmentCompatible {
+  static readonly schema = BooleanFragmentSchema;
+
+  static as(frag: FragmentCompatible): BooleanFragment {
+    if (frag instanceof BooleanFragment) {
+      return frag;
+    }
+    BooleanFragmentSchema.parse(frag);
+    Object.setPrototypeOf(frag, BooleanFragment.prototype);
+    return frag as BooleanFragment;
+  }
+
+  readonly type: FragmentType = 'boolean';
+  readonly isComplete = true;
+  readonly value: boolean;
+  constructor(input: boolean | string) {
+    this.value = typeof input === 'string' ? starlark.asBoolean(input) : input;
+  }
+}
+
+export class ArrayFragment implements FragmentCompatible {
+  static readonly schema = ArrayFragmentSchema;
+
+  static as(frag: FragmentCompatible): ArrayFragment {
+    if (frag instanceof ArrayFragment) {
+      return frag;
+    }
+    ArrayFragmentSchema.parse(frag);
+    Object.setPrototypeOf(frag, ArrayFragment.prototype);
+    const array = frag as ArrayFragment;
+    for (let i = 0; i < array.items.length; i++) {
+      array.items[i] = Fragments.asValue(array.items[i]);
+    }
+    return array;
+  }
+
+  readonly type: FragmentType = 'array';
+  isComplete = false;
+  items: ValueFragment[] = [];
+
+  constructor(items: ValueFragment[] = [], isComplete = false) {
+    this.items = items;
+    this.isComplete = isComplete;
+  }
+
+  addValue(item: ValueFragment): void {
+    this.items.push(item);
+  }
+}
+
+export class RecordFragment implements FragmentCompatible {
+  static readonly schema = RecordFragmentSchema;
+
+  static safeAs(frag: FragmentCompatible): RecordFragment | undefined {
+    if (frag instanceof RecordFragment) {
+      return frag;
+    }
+    const parseResult = RecordFragmentSchema.safeParse(frag);
+    if (!parseResult.success) {
+      return undefined;
+    }
+    Object.setPrototypeOf(frag, RecordFragment.prototype);
+    const record = frag as RecordFragment;
+    for (const prop in record.children) {
+      const child = record.children[prop];
+      record.children[prop] = Fragments.asValue(child);
+    }
+    return record;
+  }
+
+  static as(frag: FragmentCompatible): RecordFragment {
+    const record = RecordFragment.safeAs(frag);
+    if (record) {
+      return record;
+    }
+    throw Fragments.typeError('record', frag.type);
+  }
+
+  readonly type: FragmentType = 'record';
+  isComplete = false;
+  children: ChildFragments;
+
+  constructor(children: ChildFragments = {}, isComplete = false) {
+    this.children = children;
+    this.isComplete = isComplete;
+  }
+
+  addAttribute(attrib: AttributeFragment): void {
+    if (!attrib.value) {
+      throw new Error('The attribute fragment does not have a value.');
+    }
+    this.children[attrib.name] = attrib.value;
+  }
+}
+
+export class AttributeFragment implements FragmentCompatible {
+  static readonly schema = AttributeFragmentSchema;
+
+  static safeAs(frag: FragmentCompatible): AttributeFragment | undefined {
```

Comment:
```suggestion
  static safeAs(frag: FragmentCompatible): AttributeFragment | null {
```
Use null rather than undefined when actively returning values.